File: src/core/strategies/langgraph/strategy.py
import json
import logging
import time
from datetime import datetime

from langchain_core.language_models.chat_models import BaseChatModel
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langchain_core.messages.base import BaseMessage
from langgraph.checkpoint.memory import MemorySaver
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from src.core.base_strategy import BaseStrategy
from src.core.models import QueryContext, QueryResult
from src.core.strategy_registry import register_strategy
from src.core.strategies.langgraph.graph import build_fhir_graph
from src.core.strategies.langgraph.state import ConversationState
from src.core.strategies.langgraph.tools import set_run_context
from src.core.strategies.utils.prompts import SYSTEM_PROMPT

logger = logging.getLogger(__name__)


def _extract_plain_text(content: str | list | dict | None) -> str:
    """Extract plain text from Gemini block format [{'type':'text','text':'...'}] or plain str."""
    if content is None:
        return ""
    if isinstance(content, str):
        s = content.strip()
        if not s:
            return ""
        try:
            parsed = json.loads(s)
            if isinstance(parsed, list):
                parts = []
                for block in parsed:
                    if isinstance(block, dict) and "text" in block:
                        parts.append(str(block["text"]))
                    elif isinstance(block, str):
                        parts.append(block)
                return "\n".join(parts).strip() if parts else s
        except (json.JSONDecodeError, TypeError):
            pass
        return s
    if isinstance(content, list):
        parts = []
        for block in content:
            if isinstance(block, dict) and "text" in block:
                parts.append(str(block["text"]))
            elif isinstance(block, str):
                parts.append(block)
        return "\n".join(parts).strip()
    if isinstance(content, dict) and "text" in content:
        return str(content["text"]).strip()
    return str(content).strip() if content else ""


def _extract_final(messages: list[BaseMessage]) -> tuple[str, list[str]]:
    """Return (answer, resource_ids) from finish_with_answer ToolMessage or fallback to last AIMessage."""
    for msg in reversed(messages):
        if hasattr(msg, "name") and getattr(msg, "name", None) == "finish_with_answer":
            if hasattr(msg, "content") and msg.content:
                try:
                    data = json.loads(msg.content) if isinstance(msg.content, str) else msg.content
                    if isinstance(data, dict):
                        raw = data.get("answer", "")
                        answer = _extract_plain_text(raw).strip()
                        resource_ids = list(dict.fromkeys(data.get("resource_ids") or []))
                        if answer:
                            return answer, resource_ids
                except (json.JSONDecodeError, TypeError):
                    pass

    for msg in reversed(messages):
        if isinstance(msg, AIMessage) and not getattr(msg, "tool_calls", None):
            content = getattr(msg, "content", None)
            answer = _extract_plain_text(content).strip()
            if answer:
                return answer, []

    return "Sorry, I could not generate an answer.", []


@register_strategy("langgraph")
class LanggraphStrategy(BaseStrategy):
    """Thin strategy wrapper. Graph is compiled once in __init__."""

    def __init__(
        self,
        session_factory: async_sessionmaker[AsyncSession],
        llm: BaseChatModel,
        graph=None,
    ) -> None:
        self.session_factory = session_factory
        self.llm = llm
        self._graph = graph or build_fhir_graph(
            session_factory=session_factory,
            llm=llm,
            checkpointer=MemorySaver(),
        )

    @property
    def name(self) -> str:
        return "langgraph"

    async def execute(self, context: QueryContext) -> QueryResult:
        resource_types_collector: set[str] = set()
        set_run_context(context.patient_id, resource_types_collector)

        try:
            t0 = time.perf_counter()
            initial_state: ConversationState = {
                "messages": [
                    SystemMessage(
                        content=SYSTEM_PROMPT.format(
                            current_date=datetime.now().strftime("%B %d, %Y")
                        )
                    ),
                    HumanMessage(content=context.query_text),
                ],
                "patient_id": context.patient_id,
                "turn_count": 0,
                "tokens_in": 0,
                "tokens_out": 0,
            }

            config = {"configurable": {"thread_id": f"patient-{context.patient_id}"}}
            final_state = await self._graph.ainvoke(initial_state, config=config)

            messages = final_state.get("messages", [])
            answer, resource_ids = _extract_final(messages)
            latency_ms = (time.perf_counter() - t0) * 1000
            model_id = getattr(self.llm, "model", None) or "unknown"

            logger.info(
                "langgraph_complete | patient=%s | latency_ms=%.0f | resource_ids=%d | resource_types=%s",
                context.patient_id,
                latency_ms,
                len(resource_ids),
                len(resource_types_collector),
            )

            return QueryResult(
                response_text=answer,
                resource_ids=resource_ids,
                model_used=model_id,
                strategy_used=self.name,
                latency_ms=latency_ms,
                tokens_in=final_state.get("tokens_in", 0),
                tokens_out=final_state.get("tokens_out", 0),
                resource_types=sorted(resource_types_collector),
            )

        except Exception as e:
            error_message = str(e)
            if "429" in error_message or "RESOURCE_EXHAUSTED" in str(e).upper():
                error_message = "Model Rate limit exceeded. Please try again later."

            logger.error(
                "strategy_failed | strategy=%s | patient_id=%s | error=%s",
                self.name,
                context.patient_id,
                error_message,
            )

            return QueryResult(
                response_text="",
                resource_ids=[],
                error=error_message,
                resource_types=[],
                model_used=getattr(self.llm, "model", None) or "unknown",
                strategy_used=self.name,
            )

File: ts-version/src/core/strategies/langgraph/strategy.ts
import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { HumanMessage, SystemMessage, AIMessage } from '@langchain/core/messages';

import { BaseStrategy, QueryContext, QueryResult } from '../../models';
import { ConversationState, StreamEvent } from './state';
import { buildFHIRGraph } from './graph';
import { setRunContext } from './tools';
import { SYSTEM_PROMPT } from '../utils/prompts';
import { logger } from '../../../config/logging';
import { registerStrategy } from '../../strategy-registry';

@registerStrategy('langgraph')
export class LanggraphStrategy implements BaseStrategy {
  readonly name = 'langgraph';

  /**
   * Initialize LangGraph strategy with standard LangChain components.
   * 
   * @param dbPool - Database connection pool
   * @param llm - Standard LangChain BaseChatModel (no custom wrapper needed)
   */
  constructor(
    private dbPool: any,
    private llm: BaseChatModel
  ) {
    // Build graph using standard LangChain integration
    this._graph = buildFHIRGraph(this.dbPool, this.llm);
  }

  private _graph: any;

  async executeWithStreaming(context: QueryContext, onEvent: (event: StreamEvent) => void): Promise<QueryResult> {
    const resourceTypesCollector: Set<string> = new Set();
    setRunContext(context.patientId, resourceTypesCollector);

    try {
      const startTime = Date.now();
      
      // Emit thinking event
      onEvent({
        type: 'thinking',
        data: { message: 'Analyzing your health question...' },
        timestamp: new Date().toISOString()
      });
      
      // Format system prompt with current date
      const currentDate = new Date().toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'long',
        day: 'numeric'
      });
      const formattedPrompt = SYSTEM_PROMPT.replace('{current_date}', currentDate);
      
      // ✅ Check if conversation exists
      const existingState = await this._graph.getState({ 
        configurable: { thread_id: `patient-${context.patientId}` } 
      });
      
      const hasExistingConversation = existingState?.values?.messages?.length > 0;
      
      // ✅ Fresh conversation: system message + human message
      // ✅ Existing conversation: just human message (graph has the rest)
      const initialMessages = hasExistingConversation
        ? [new HumanMessage(context.queryText)]
        : [new SystemMessage(formattedPrompt), new HumanMessage(context.queryText)];

      const initialState: ConversationState = {
        messages: initialMessages,
        patientId: context.patientId,
        turnCount: 0,
        tokensIn: 0,
        tokensOut: 0,
        onEvent, // Pass callback through state
      };

      const config = { configurable: { thread_id: `patient-${context.patientId}` } };
      
      // Use graph streaming for real-time events
      let finalState;
      for await (const chunk of this._graph.stream(initialState, config)) {
        finalState = chunk.state;
        
        // Emit graph-level events
        if (chunk.node === 'llm') {
          onEvent({
            type: 'graph_step',
            data: {
              node: 'llm',
              message: 'AI is analyzing and deciding next actions...',
              turnCount: chunk.state.turnCount,
              tokensSoFar: (chunk.state.tokensIn || 0) + (chunk.state.tokensOut || 0)
            },
            timestamp: new Date().toISOString()
          });
        } else if (chunk.node === 'tools') {
          const lastMessage = chunk.state.messages[chunk.state.messages.length - 1];
          const toolCalls = (lastMessage as any)?.tool_calls || [];
          
          onEvent({
            type: 'graph_step',
            data: {
              node: 'tools',
              message: 'Executing health data retrieval...',
              toolCount: toolCalls.length
            },
            timestamp: new Date().toISOString()
          });
        }
        
        // ToolNode wrapper will emit tool-level events when onEvent is present
        // No additional events needed here since ToolNode handles it
      }

      const { answer, resourceIds } = this.extractFinal(finalState.messages || []);
      const latencyMs = Date.now() - startTime;
      const modelId = (this.llm as any).model || 'unknown';

      // Emit completion event
      onEvent({
        type: 'complete',
        data: {
          finalResponse: answer,
          resourceIds,
          tokensIn: finalState.tokensIn || 0,
          tokensOut: finalState.tokensOut || 0,
          latencyMs,
          modelUsed: modelId,
          strategyUsed: this.name,
          resourceTypes: Array.from(resourceTypesCollector).sort()
        },
        timestamp: new Date().toISOString()
      });

      logger.info(
        `langgraph_complete | patient=${context.patientId} | latency_ms=${latencyMs} | resource_ids=${resourceIds.length} | resource_types=${resourceTypesCollector.size}`,
        { latencyMs, patientId: context.patientId, resourceCount: resourceIds.length, resourceTypeCount: resourceTypesCollector.size }
      );

      return {
        responseText: answer,
        resourceIds,
        modelUsed: modelId,
        strategyUsed: this.name,
        tokensIn: finalState.tokensIn || 0,
        tokensOut: finalState.tokensOut || 0,
        latencyMs,
        resourceTypes: Array.from(resourceTypesCollector).sort(),
      };

    } catch (error) {
      let errorMessage = String(error);
      if (errorMessage.includes('429') || errorMessage.toUpperCase().includes('RESOURCE_EXHAUSTED')) {
        errorMessage = 'Model Rate limit exceeded. Please try again later.';
      }

      // Emit error event
      onEvent({
        type: 'error',
        data: {
          error: 'Processing failed',
          message: errorMessage
        },
        timestamp: new Date().toISOString()
      });

      logger.error(
        `strategy_failed | strategy=${this.name} | patient_id=${context.patientId} | error=${errorMessage}`,
        { error, strategy: this.name, patientId: context.patientId }
      );

      return {
        responseText: '',
        resourceIds: [],
        modelUsed: (this.llm as any).model || 'unknown',
        strategyUsed: this.name,
        tokensIn: 0,
        tokensOut: 0,
        latencyMs: 0,
        error: errorMessage,
        resourceTypes: [],
      };
    }
  }

  async execute(context: QueryContext): Promise<QueryResult> {
    const resourceTypesCollector: Set<string> = new Set();
    setRunContext(context.patientId, resourceTypesCollector);

    try {
      const startTime = Date.now();
      
      // Format system prompt with current date
      const currentDate = new Date().toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'long',
        day: 'numeric'
      });
      const formattedPrompt = SYSTEM_PROMPT.replace('{current_date}', currentDate);
      
      // ✅ Check if conversation exists
      const existingState = await this._graph.getState({ 
        configurable: { thread_id: `patient-${context.patientId}` } 
      });
      
      const hasExistingConversation = existingState?.values?.messages?.length > 0;
      
      // ✅ Fresh conversation: system message + human message
      // ✅ Existing conversation: just human message (graph has the rest)
      const initialMessages = hasExistingConversation
        ? [new HumanMessage(context.queryText)]
        : [new SystemMessage(formattedPrompt), new HumanMessage(context.queryText)];

      const initialState: ConversationState = {
        messages: initialMessages,
        patientId: context.patientId,
        turnCount: 0,
        tokensIn: 0,
        tokensOut: 0,
      };

      const config = { configurable: { thread_id: `patient-${context.patientId}` } };
      const finalState = await this._graph.invoke(initialState, config);

      const { answer, resourceIds } = this.extractFinal(finalState.messages || []);
      const latencyMs = Date.now() - startTime;
      const modelId = (this.llm as any).model || 'unknown';

      logger.info(
        `langgraph_complete | patient=${context.patientId} | latency_ms=${latencyMs} | resource_ids=${resourceIds.length} | resource_types=${resourceTypesCollector.size}`,
        { latencyMs, patientId: context.patientId, resourceCount: resourceIds.length, resourceTypeCount: resourceTypesCollector.size }
      );

      return {
        responseText: answer,
        resourceIds,
        modelUsed: modelId,
        strategyUsed: this.name,
        tokensIn: finalState.tokensIn || 0,
        tokensOut: finalState.tokensOut || 0,
        latencyMs,
        resourceTypes: Array.from(resourceTypesCollector).sort(),
      };

    } catch (error) {
      let errorMessage = String(error);
      if (errorMessage.includes('429') || errorMessage.toUpperCase().includes('RESOURCE_EXHAUSTED')) {
        errorMessage = 'Model Rate limit exceeded. Please try again later.';
      }

      logger.error(
        `strategy_failed | strategy=${this.name} | patient_id=${context.patientId} | error=${errorMessage}`,
        { error, strategy: this.name, patientId: context.patientId }
      );

      return {
        responseText: '',
        resourceIds: [],
        modelUsed: (this.llm as any).model || 'unknown',
        strategyUsed: this.name,
        tokensIn: 0,
        tokensOut: 0,
        latencyMs: 0,
        error: errorMessage,
        resourceTypes: [],
      };
    }
  }

  private extractFinal(messages: any[]): { answer: string; resourceIds: string[] } {
    // Look for finish_with_answer ToolMessage first
    for (const msg of messages.reverse()) {
      if ((msg as any).name === 'finish_with_answer' && (msg as any).content) {
        try {
          const data = JSON.parse((msg as any).content);
          if (typeof data === 'object' && data !== null) {
            const answer = this.extractPlainText(data.answer);
            const resourceIds = Array.from(new Set(data.resource_ids || [])) as string[];
            if (answer) {
              return { answer, resourceIds };
            }
          }
        } catch {
          // Ignore JSON parse errors
        }
      }
    }

    // Fallback to last AIMessage without tool calls
    for (const msg of messages.reverse()) {
      if (msg instanceof AIMessage && !(msg.tool_calls && msg.tool_calls.length > 0)) {
        const answer = this.extractPlainText(msg.content);
        if (answer) {
          return { answer, resourceIds: [] };
        }
      }
    }

    return { answer: 'Sorry, I could not generate an answer.', resourceIds: [] };
  }

  private extractPlainText(content: any): string {
    if (content === null) return '';
    if (typeof content === 'string') {
      const trimmed = content.trim();
      if (!trimmed) return '';
      
      try {
        const parsed = JSON.parse(trimmed);
        if (Array.isArray(parsed)) {
          const parts: string[] = [];
          for (const block of parsed) {
            if (typeof block === 'object' && block !== null && 'text' in block) {
              parts.push(String(block.text));
            } else if (typeof block === 'string') {
              parts.push(block);
            }
          }
          return parts.join('\n').trim() || trimmed;
        }
      } catch {
        // Not JSON, return as-is
      }
      return trimmed;
    }
    
    if (Array.isArray(content)) {
      const parts: string[] = [];
      for (const block of content) {
        if (typeof block === 'object' && block !== null && 'text' in block) {
          parts.push(String(block.text));
        } else if (typeof block === 'string') {
          parts.push(block);
        }
      }
      return parts.join('\n').trim();
    }
    
    if (typeof content === 'object' && content !== null && 'text' in content) {
      return String(content.text).trim();
    }
    
    return String(content).trim() || '';
  }
}
